seed/analysis_pipelines/eui.py

```diff
@@ -59,73 +62,83 @@ def _get_valid_meters(property_view_ids):
+        ).order_by('start_time'):
+            if meter_reading.meter.id not in meter_readings_by_meter:
+                meter_readings_by_meter[meter_reading.meter.id] = []
+            meter_readings_by_meter[meter_reading.meter.id].append(meter_reading)
+
+        # generate summary per meter
+        done = False
+        readings_by_meter = {}
+        for meter_id in meter_readings_by_meter:
+            last_reading = None
+            total_time = 0
+            total_reading = 0
+            for reading in meter_readings_by_meter[meter_id]:
+
+                # ensure no overlapping readings per meter
+                if last_reading is not None:
+                    if last_reading.end_time > reading.start_time:
+                        overlapping_meter.append(property_view.id)
+                        done = True
+                        break
+
+                last_reading = reading
+                total_time += (reading.end_time - reading.start_time).total_seconds()
+                total_reading += reading.reading
+            if done:
+                continue
+            readings_by_meter[meter_id] = {'time': total_time, 'reading': total_reading}
+
+        # done with this property_view
+        if readings_by_meter:
+            meter_readings_by_property_view[property_view.id] = readings_by_meter
 
     errors_by_property_view_id = {}
     for pid in invalid_area:
         if pid not in errors_by_property_view_id:
             errors_by_property_view_id[pid] = []
         errors_by_property_view_id[pid].append(EUI_ANALYSIS_MESSAGES[ERROR_INVALID_GROSS_FLOOR_AREA])
-    for pid in invalid_meter_1:
+    for pid in invalid_meter:
         if pid not in errors_by_property_view_id:
             errors_by_property_view_id[pid] = []
-        errors_by_property_view_id[pid].append(EUI_ANALYSIS_MESSAGES[ERROR_INSUFFICIENT_METER_READINGS])
-    for pid in invalid_meter_2:
+        errors_by_property_view_id[pid].append(EUI_ANALYSIS_MESSAGES[ERROR_INVALID_METER_READINGS])
+    for pid in overlapping_meter:
         if pid not in errors_by_property_view_id:
             errors_by_property_view_id[pid] = []
-        errors_by_property_view_id[pid].append(EUI_ANALYSIS_MESSAGES[ERROR_INVALID_METER_READINGS])
+        errors_by_property_view_id[pid].append(EUI_ANALYSIS_MESSAGES[ERROR_OVERLAPPING_METER_READINGS])
 
     return meter_readings_by_property_view, errors_by_property_view_id
 
 
 def _calculate_eui(meter_readings, gross_floor_area):
-    return round(sum(meter_readings) / gross_floor_area, 4)
+    total_reading = 0
+    total_time = 0
+    for meter_id in meter_readings:
+        total_reading += meter_readings[meter_id]['reading']
+        total_time += meter_readings[meter_id]['time']
+    return {
+        'eui': round(total_reading / gross_floor_area, 4),
+        'reading': total_reading,
+        'coverage': 100 - round(100 * (TIME_PERIOD.total_seconds() - total_time) / TIME_PERIOD.total_seconds())
+    }
```

Comment:
I think I might have figured out why we saw > 100% for coverage on some properties, there could be two meters which report overlapping reading intervals, but wouldn't get filtered out by the "overlapping_meters" check above b/c they're separate meters
![Screen Shot 2021-10-04 at 4 29 17 PM](https://user-images.githubusercontent.com/18518728/135933302-ee2e5e49-6277-4e21-b153-fa4c0a86cdb1.png)

I think including both of these meters and their readings is actually OK -- a building could have multiple separate meters. The tricky part now is handling the total time...

Maybe the simplest approach could be to keep a hash of datetime days (at midnight or something) from the meter readings. Then count the number of keys `len(meter_days_hash.keys())`... not perfectly accurate but pretty good and not too resource intensive/complex

Alternatively we could sort _all_ readings, and add up the diffs for non-overlapping days. Interested to hear what you think though